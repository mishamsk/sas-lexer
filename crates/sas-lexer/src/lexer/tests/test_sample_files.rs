use super::{
    super::{lex_program, LexResult},
    util::{error_to_string, token_to_string},
};
use insta::assert_yaml_snapshot;
use rstest::rstest;
use std::{fs, path::PathBuf};

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

#[rstest]
fn test_snapshots(#[files("src/lexer/tests/samples/**/*.sas")] path: PathBuf) {
    // Compute the absolute path of the prefix

    let prefix = fs::canonicalize(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/src/lexer/tests/samples/"
    ))
    .unwrap();

    let snap_name = path.strip_prefix(&prefix).unwrap();
    let snap_name_str = snap_name
        .to_str()
        .unwrap()
        .replace(std::path::MAIN_SEPARATOR, "__");

    set_snapshot_suffix!("{}", snap_name_str);

    let contents = fs::read_to_string(&path).unwrap();
    let LexResult {
        buffer: tok_buffer,
        errors,
        ..
    } = lex_program(&contents).unwrap();
    let tokens: Vec<String> = tok_buffer
        .into_iter()
        .map(|tidx| token_to_string(tidx, &tok_buffer, &contents))
        .collect();

    assert_yaml_snapshot!("tokens", tokens);

    if errors.is_empty() {
        insta::allow_duplicates! {
            assert_yaml_snapshot!("errors_snapshot", @r###"
            errors_snapshot
            "###);
        }
    } else {
        let error_strings = errors
            .iter()
            .map(|error| error_to_string(error, &tok_buffer, &contents).to_string())
            .collect::<Vec<String>>();
        assert_yaml_snapshot!("errors_snapshot", error_strings);
    }
}

/// Tests that tokens cover all offsets in the file from 0 to the end
#[rstest]
fn test_full_coverage(#[files("src/lexer/tests/samples/**/*.sas")] path: PathBuf) {
    let contents = fs::read_to_string(&path).unwrap();

    let LexResult {
        buffer: tok_buffer, ..
    } = lex_program(&contents).unwrap();

    let mut end = 0;

    for token in &tok_buffer {
        // Check that the token starts where the previous token ended
        assert_eq!(
            tok_buffer
                .get_token_start(token)
                .expect("wrong token")
                .get(),
            end,
            "Token <{}> does not start where the previous token ended",
            token_to_string(token, &tok_buffer, &contents)
        );

        // Set the new end
        end = tok_buffer.get_token_end(token).expect("wrong token").get();

        // Check that the end is greater than the start
        assert!(
            end >= tok_buffer
                .get_token_start(token)
                .expect("wrong token")
                .get(),
            "Token <{}> has an end before the start",
            token_to_string(token, &tok_buffer, &contents)
        );
    }

    // Check that the last token ends at the end of the file
    assert_eq!(end, contents.len() as u32);
}
