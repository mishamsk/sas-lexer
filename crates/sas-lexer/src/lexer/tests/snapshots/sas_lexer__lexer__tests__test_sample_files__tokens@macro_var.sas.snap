---
source: crates/sas-lexer/src/lexer/tests/test_sample_files.rs
expression: tokens
snapshot_kind: text
---
- "[@0,0:1=\"&\",<AMP>,L1:C0-L1:C1,chl=<D>,pl=<None>]"
- "[@1,1:2=\"\n\",<WS>,L1:C1-L1:C2,chl=<H>,pl=<None>]"
- "[@2,2:6=\"&&&&\",<AMP>,L2:C0-L2:C4,chl=<D>,pl=<None>]"
- "[@3,6:7=\"\n\",<WS>,L2:C4-L2:C5,chl=<H>,pl=<None>]"
- "[@4,7:8=\"&\",<AMP>,L3:C0-L3:C1,chl=<D>,pl=<None>]"
- "[@5,8:9=\"9\",<IntegerLiteral>,L3:C1-L3:C2,chl=<D>,pl=9]"
- "[@6,9:10=\"\n\",<WS>,L3:C2-L3:C3,chl=<H>,pl=<None>]"
- "[@7,10:15=\"&mvar\",<MacroVarExpr>,L4:C0-L4:C5,chl=<D>,pl=<None>]"
- "[@8,15:16=\"\n\",<WS>,L4:C5-L4:C6,chl=<H>,pl=<None>]"
- "[@9,16:22=\"&mvar.\",<MacroVarExpr>,L5:C0-L5:C6,chl=<D>,pl=<None>]"
- "[@10,22:23=\"\n\",<WS>,L5:C6-L5:C7,chl=<H>,pl=<None>]"
- "[@11,23:28=\"&mvar\",<MacroVarExpr>,L6:C0-L6:C5,chl=<D>,pl=<None>]"
- "[@12,28:29=\"&\",<AMP>,L6:C5-L6:C6,chl=<D>,pl=<None>]"
- "[@13,29:30=\"\n\",<WS>,L6:C6-L6:C7,chl=<H>,pl=<None>]"
- "[@14,30:36=\"&mvar.\",<MacroVarExpr>,L7:C0-L7:C6,chl=<D>,pl=<None>]"
- "[@15,36:37=\"&\",<AMP>,L7:C6-L7:C7,chl=<D>,pl=<None>]"
- "[@16,37:38=\"\n\",<WS>,L7:C7-L7:C8,chl=<H>,pl=<None>]"
- "[@17,38:43=\"&mvar\",<MacroVarExpr>,L8:C0-L8:C5,chl=<D>,pl=<None>]"
- "[@18,43:44=\"&\",<AMP>,L8:C5-L8:C6,chl=<D>,pl=<None>]"
- "[@19,44:45=\".\",<DOT>,L8:C6-L8:C7,chl=<D>,pl=<None>]"
- "[@20,45:46=\"\n\",<WS>,L8:C7-L8:C8,chl=<H>,pl=<None>]"
- "[@21,46:52=\"&&mvar\",<MacroVarExpr>,L9:C0-L9:C6,chl=<D>,pl=<None>]"
- "[@22,52:53=\"\n\",<WS>,L9:C6-L9:C7,chl=<H>,pl=<None>]"
- "[@23,53:60=\"&&mvar.\",<MacroVarExpr>,L10:C0-L10:C7,chl=<D>,pl=<None>]"
- "[@24,60:61=\"\n\",<WS>,L10:C7-L10:C8,chl=<H>,pl=<None>]"
- "[@25,61:67=\"&&mvar\",<MacroVarExpr>,L11:C0-L11:C6,chl=<D>,pl=<None>]"
- "[@26,67:68=\"&\",<AMP>,L11:C6-L11:C7,chl=<D>,pl=<None>]"
- "[@27,68:69=\"\n\",<WS>,L11:C7-L11:C8,chl=<H>,pl=<None>]"
- "[@28,69:76=\"&&mvar.\",<MacroVarExpr>,L12:C0-L12:C7,chl=<D>,pl=<None>]"
- "[@29,76:77=\"&\",<AMP>,L12:C7-L12:C8,chl=<D>,pl=<None>]"
- "[@30,77:78=\"\n\",<WS>,L12:C8-L12:C9,chl=<H>,pl=<None>]"
- "[@31,78:84=\"&&mvar\",<MacroVarExpr>,L13:C0-L13:C6,chl=<D>,pl=<None>]"
- "[@32,84:85=\"&\",<AMP>,L13:C6-L13:C7,chl=<D>,pl=<None>]"
- "[@33,85:86=\"9\",<IntegerLiteral>,L13:C7-L13:C8,chl=<D>,pl=9]"
- "[@34,86:87=\"\n\",<WS>,L13:C8-L13:C9,chl=<H>,pl=<None>]"
- "[@35,87:94=\"&&mvar.\",<MacroVarExpr>,L14:C0-L14:C7,chl=<D>,pl=<None>]"
- "[@36,94:95=\"&\",<AMP>,L14:C7-L14:C8,chl=<D>,pl=<None>]"
- "[@37,95:96=\"9\",<IntegerLiteral>,L14:C8-L14:C9,chl=<D>,pl=9]"
- "[@38,96:97=\"\n\",<WS>,L14:C9-L14:C10,chl=<H>,pl=<None>]"
- "[@39,97:108=\"&mvar&mvar2\",<MacroVarExpr>,L15:C0-L15:C11,chl=<D>,pl=<None>]"
- "[@40,108:109=\"\n\",<WS>,L15:C11-L15:C12,chl=<H>,pl=<None>]"
- "[@41,109:121=\"&mvar&mvar2.\",<MacroVarExpr>,L16:C0-L16:C12,chl=<D>,pl=<None>]"
- "[@42,121:122=\"\n\",<WS>,L16:C12-L16:C13,chl=<H>,pl=<None>]"
- "[@43,122:128=\"&mvar.\",<MacroVarExpr>,L17:C0-L17:C6,chl=<D>,pl=<None>]"
- "[@44,128:134=\"&mvar2\",<MacroVarExpr>,L17:C6-L17:C12,chl=<D>,pl=<None>]"
- "[@45,134:135=\"\n\",<WS>,L17:C12-L17:C13,chl=<H>,pl=<None>]"
- "[@46,135:141=\"&mvar.\",<MacroVarExpr>,L18:C0-L18:C6,chl=<D>,pl=<None>]"
- "[@47,141:148=\"&mvar2.\",<MacroVarExpr>,L18:C6-L18:C13,chl=<D>,pl=<None>]"
- "[@48,148:149=\"\n\",<WS>,L18:C13-L18:C14,chl=<H>,pl=<None>]"
- "[@49,149:161=\"&&mvar&mvar2\",<MacroVarExpr>,L19:C0-L19:C12,chl=<D>,pl=<None>]"
- "[@50,161:162=\"\n\",<WS>,L19:C12-L19:C13,chl=<H>,pl=<None>]"
- "[@51,162:175=\"&&mvar&mvar2.\",<MacroVarExpr>,L20:C0-L20:C13,chl=<D>,pl=<None>]"
- "[@52,175:176=\"\n\",<WS>,L20:C13-L20:C14,chl=<H>,pl=<None>]"
- "[@53,176:183=\"&&mvar.\",<MacroVarExpr>,L21:C0-L21:C7,chl=<D>,pl=<None>]"
- "[@54,183:189=\"&mvar2\",<MacroVarExpr>,L21:C7-L21:C13,chl=<D>,pl=<None>]"
- "[@55,189:190=\"\n\",<WS>,L21:C13-L21:C14,chl=<H>,pl=<None>]"
- "[@56,190:197=\"&&mvar.\",<MacroVarExpr>,L22:C0-L22:C7,chl=<D>,pl=<None>]"
- "[@57,197:204=\"&mvar2.\",<MacroVarExpr>,L22:C7-L22:C14,chl=<D>,pl=<None>]"
- "[@58,204:205=\"\n\",<WS>,L22:C14-L22:C15,chl=<H>,pl=<None>]"
- "[@59,205:218=\"&&&mvar&mvar2\",<MacroVarExpr>,L23:C0-L23:C13,chl=<D>,pl=<None>]"
- "[@60,218:219=\"\n\",<WS>,L23:C13-L23:C14,chl=<H>,pl=<None>]"
- "[@61,219:233=\"&&&mvar&mvar2.\",<MacroVarExpr>,L24:C0-L24:C14,chl=<D>,pl=<None>]"
- "[@62,233:234=\"\n\",<WS>,L24:C14-L24:C15,chl=<H>,pl=<None>]"
- "[@63,234:242=\"&&&mvar.\",<MacroVarExpr>,L25:C0-L25:C8,chl=<D>,pl=<None>]"
- "[@64,242:248=\"&mvar2\",<MacroVarExpr>,L25:C8-L25:C14,chl=<D>,pl=<None>]"
- "[@65,248:249=\"\n\",<WS>,L25:C14-L25:C15,chl=<H>,pl=<None>]"
- "[@66,249:257=\"&&&mvar.\",<MacroVarExpr>,L26:C0-L26:C8,chl=<D>,pl=<None>]"
- "[@67,257:264=\"&mvar2.\",<MacroVarExpr>,L26:C8-L26:C15,chl=<D>,pl=<None>]"
- "[@68,264:265=\"\n\",<WS>,L26:C15-L26:C16,chl=<H>,pl=<None>]"
- "[@69,265:265=\"<no text>\",<EOF>,L27:C0-L27:C0,chl=<D>,pl=<None>]"
